server/base/src/main/java/org/apache/accumulo/server/conf/store/PropCacheId.java

```diff
@@ -0,0 +1,278 @@
+   * the final token is PROP_NODE_NAME and then the type is defined if the path has table or
+   * namespace in the path, otherwise it is assumed to be system.
+   *
+   * @param tokens
+   *          a path split into String[] of tokens
+   * @return the id type.
+   */
+  public static IdType extractType(final String[] tokens) {
+    if (tokens.length == 0 || !tokens[tokens.length - 1].equals(PROP_NODE_NAME)) {
+      // without tokens or it does not end with PROP_NAME_NAME
+      return IdType.UNKNOWN;
+    }
+    if (tokens[TYPE_TOKEN_POSITION].equals(TABLES_NODE_NAME)) {
+      return IdType.TABLE;
+    }
+    if (tokens[TYPE_TOKEN_POSITION].equals(NAMESPACE_NODE_NAME)) {
+      return IdType.NAMESPACE;
+    }
+    return IdType.SYSTEM;
+  }
+
+  public String getPath() {
+    return path;
+  }
+
+  public IdType getIdType() {
+    return idType;
+  }
+
+  @Override
+  public int compareTo(@NonNull PropCacheId other) {
+    return Comparator.comparing(PropCacheId::getIdType).thenComparing(PropCacheId::getPath)
+        .compare(this, other);
+  }
+
+  // TODO - remove optional and return null.
+  /**
+   * If the prop cache is for a namespace, return the namespace id.
+   *
+   * @return the namespace id.
+   */
+  public Optional<NamespaceId> getNamespaceId() {
+    return Optional.ofNullable(namespaceId);
+  }
+
+  /**
+   * if the prop cache is for a table, return the table id.
+   *
+   * @return the table id.
+   */
+  public Optional<TableId> getTableId() {
+    return Optional.ofNullable(tableId);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o)
+      return true;
+    if (o == null || getClass() != o.getClass())
+      return false;
+    PropCacheId that = (PropCacheId) o;
+    return path.equals(that.path);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(path);
+  }
+
+  @Override
+  public String toString() {
+    switch (idType) {
+      case SYSTEM:
+        return new StringJoiner(", ", PropCacheId.class.getSimpleName() + "[", "]")
+            .add("idType=System").toString();
+      case NAMESPACE:
+        return new StringJoiner(", ", PropCacheId.class.getSimpleName() + "[", "]")
+            .add("idType=Namespace").add("namespaceId=" + namespaceId).toString();
+      case TABLE:
+        return new StringJoiner(", ", PropCacheId.class.getSimpleName() + "[", "]")
+            .add("idType=Table").add("tableId=" + tableId).toString();
+      default:
+        return new StringJoiner(", ", PropCacheId.class.getSimpleName() + "[", "]")
+            .add("idType=" + idType).add("namespaceId=" + namespaceId).add("tableId=" + tableId)
+            .add("path='" + path + "'").toString();
+    }
+
+  }
+
+  /**
+   * Define types of properties stored in zookeeper. Note: default properties are not in zookeeper
+   * but come from code.
+   */
+  public enum IdType {
+    UNKNOWN, SYSTEM, NAMESPACE, TABLE
+  }
```

Comment:
The intent was to separate the path and the ZooKeeper structure from the PropCacheKey. It is important to know the type, but the path is an implementation detail.  The ZooKeeper storage internals may have leaked over time, but the original intent was that things using a prop store would not know that the storage was actually ZooKeeper.  In place where this abstraction has failed / leaked  my position would be to fix the leak.

I may have failed in this, but it was my original intent and warrants additional discussion if there are benefits from the abstraction, but it seems to me that it would provide benefits to "hide" ZooKeeper abstractions where possible.

Reply:
I still think there's room to clean up a bit of this, and get rid of the enum. I'll try to tackle it as a PR against your branch if I can come up with something.